src/lib.rs

```diff
@@ -156,12 +261,60 @@ fn make_original_shards(original_count: u16, data: &[u8]) -> Vec<Vec<u8>> {
 	let shard_bytes = shard_bytes(original_count, data.len());
 	assert_ne!(shard_bytes, 0);
 
-	let mut result = vec![vec![0u8; shard_bytes]; original_count as usize];
-	for (i, chunk) in data.chunks(shard_bytes).enumerate() {
-		result[i][..chunk.len()].as_mut().copy_from_slice(chunk);
+	#[cfg(not(feature = "parallel"))]
+	{
+		// Optimized sequential version with prefetch
+		let mut result = Vec::with_capacity(original_count as usize);
+		let mut remaining_data = data;
+		
+		for i in 0..original_count as usize {
+			let mut chunk = vec![0u8; shard_bytes];
+			let copy_len = remaining_data.len().min(shard_bytes);
+			
+			// Prefetch next data block
+			#[cfg(any(target_arch = "x86_64", target_arch = "x86"))]
+			if i + 1 < original_count as usize && remaining_data.len() > shard_bytes {
+				unsafe {
+					let next_ptr = remaining_data.as_ptr().add(shard_bytes);
+					if (remaining_data.len() - shard_bytes) >= 64 {
+						_mm_prefetch(next_ptr as *const i8, _MM_HINT_T0);
+					}
+				}
+			}
+			
+			// Optimized copying (compiler uses memcpy/SIMD)
+			chunk[..copy_len].copy_from_slice(&remaining_data[..copy_len]);
+			
+			if likely(remaining_data.len() >= shard_bytes) {
+				remaining_data = &remaining_data[shard_bytes..];
+			} else {
+				remaining_data = &[];
+			}
+			
+			result.push(chunk);
+		}
+		result
 	}
 
-	result
+	#[cfg(feature = "parallel")]
+	{
+		// Parallel version: create shards in parallel
+		(0..original_count as usize)
+			.into_par_iter()
```

Comment:
I undestand we want to speed it up, but this is a bit unsafe in the context of polkadot-sdk because all this threads will eat CPU from other important task, could we at least bound it to a fix number of threads if it is not already.